### ISSUE TITLE
Non-deterministic behavior in cv2.distanceTransform()

### ISSUE THREAD

**Issue by SherifN** [opencv/opencv#24082]
### System Information

OpenCV python version: 4.5.5
Operating System: Ubuntu 16.04
Python version: 3.9.15

### Detailed description

I am using cv2.distanceTransform in a deep learning application and have noticed non-deterministic training results in pytorch lightning. After setting manual seeds for all random number generators in my setup, the training was still not deterministic so I decided to have a closer look at the used data. I use cv2.distanceTransform in the ground truth data generation in my training and have noticed variation in the output using the same input.

It's possible to mitigate the problem by rounding the results to one decimal, although this is not the best solution ...
```python
result = np.round(cv2.distanceTransform(feat_map, distanceType=cv2.DIST_L2, maskSize=cv2.DIST_MASK_PRECISE), decimals=1)
```
Is there a better way to fix this problem?

### Steps to reproduce

```python
import numpy as np
import cv2

num_tests = 100
shape = (800, 800)
for i in range(num_tests):
    feat_map = np.full(shape=shape, fill_value=255, dtype=np.uint8)
    # Creating a random line
    random_x_start = np.random.randint(0, 399)
    random_x_end = np.random.randint(400, 799)
    random_y_start = np.random.randint(0, 399)
    random_y_end = np.random.randint(400, 799)
    X, Y = np.mgrid[random_x_start:random_x_end:1, random_y_start:random_y_end:1]
    line = np.vstack((X.flatten(), Y.flatten())).transpose().reshape(-1, 1, 2)
    cv2.polylines(feat_map, [line], isClosed=False, color=0)
    result_1 = cv2.distanceTransform(feat_map, distanceType=cv2.DIST_L2, maskSize=cv2.DIST_MASK_PRECISE)
    result_2 = cv2.distanceTransform(feat_map, distanceType=cv2.DIST_L2, maskSize=cv2.DIST_MASK_PRECISE)
    np.testing.assert_array_equal(result_1, result_2)
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [ ] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by dkurt** [comment]
Cannot reproduce the problem with `opencv-python==4.5.5.62` and `opencv-python==4.5.5.64`. Also tried latest 4.x branch with disabled IPP (as there is IPP code in https://github.com/lamm45/opencv/blob/4.x/modules/imgproc/src/distransform.cpp).

Do you still have problem on the latest 4.8.0 release? Can you share your CPU arch and details from `cat /proc/cpuinfo`?

**Comment by a-mos** [comment]
> Cannot reproduce the problem with `opencv-python==4.5.5.62` and `opencv-python==4.5.5.64`. Also tried latest 4.x branch with disabled IPP (as there is IPP code in https://github.com/opencv/opencv/blob/4.x/modules/imgproc/src/distransform.cpp).
> 
> Do you still have problem on the latest 4.8.0 release? Can you share your CPU arch and details from `cat /proc/cpuinfo`?


Got same problem when fix number of threads to 1 (cv2.setNumThreads(1)).
Using all threads, current (4.8.0) implementation return deterministic results.

![image](https://github.com/opencv/opencv/assets/55856406/063baabb-edd2-490a-97ee-0a7bf2da8d77)

I use AMD Ryzen Threadripper 1900X CPU

**Comment by dkurt** [comment]
@a-mos, thanks! Can confirm an error on 4.x branch with IPP enabled (WITH_IPP=OFF works fine).

**Comment by dkurt** [comment]
@opencv-alalek, @mshabunin, @eplankin, can you recommend how to assign the bug to IPP team? C++ reproducer:
```cpp
int cols = 100, rows = 100;
std::vector<Ipp32f> ref(cols * rows);
std::vector<Ipp32f> dst(cols * rows);
for (int i = 0; i < 2; ++i) {
    std::vector<Ipp8u> src(cols * rows, 255);
    for (int y = 32; y < 49; ++y) {
        for (int x = 36; x < 78; ++x) {
            src[y * cols + x] = 0;
        }
    }

    IppStatus status;
    IppiSize roi = { cols, rows };
    Ipp8u *pBuffer;
    int bufSize=0;

    status = ippiTrueDistanceTransformGetBufferSize_8u32f_C1R(roi, &bufSize);

    pBuffer = (Ipp8u *)ippMalloc_L( bufSize );
    if (i == 0) {
        status = ippiTrueDistanceTransform_8u32f_C1R(src.data(), cols, ref.data(), cols*sizeof(float), roi, pBuffer);
    } else {
        status = ippiTrueDistanceTransform_8u32f_C1R(src.data(), cols, dst.data(), cols*sizeof(float), roi, pBuffer);
    }
    ippFree( pBuffer );
}

for (int y = 0; y < rows; ++y) {
    for (int x = 0; x < cols; ++x) {
        if (ref[y * cols + x] != dst[y * cols + x]) {
            printf("(x=%d, y=%d) diff %e\n", x, y, fabs(ref[y * cols + x] - dst[y * cols + x]));
        }
    }
}
```
```
(x=1, y=0) diff 3.814697e-06
(x=2, y=0) diff 3.814697e-06
(x=99, y=0) diff 3.814697e-06
(x=1, y=1) diff 3.814697e-06
(x=98, y=1) diff 3.814697e-06
(x=2, y=3) diff 3.814697e-06
(x=3, y=4) diff 3.814697e-06
(x=0, y=5) diff 3.814697e-06
(x=2, y=5) diff 3.814697e-06
(x=3, y=5) diff 3.814697e-06
(x=1, y=6) diff 3.814697e-06
(x=97, y=6) diff 3.814697e-06
(x=99, y=6) diff 3.814697e-06
(x=2, y=8) diff 3.814697e-06
(x=0, y=9) diff 3.814697e-06
(x=3, y=9) diff 3.814697e-06
(x=98, y=10) diff 3.814697e-06
(x=99, y=11) diff 3.814697e-06
(x=0, y=13) diff 3.814697e-06
(x=2, y=14) diff 3.814697e-06
(x=3, y=16) diff 3.814697e-06
(x=96, y=16) diff 1.907349e-06
(x=1, y=19) diff 3.814697e-06
(x=96, y=22) diff 1.907349e-06
(x=1, y=23) diff 3.814697e-06
(x=99, y=23) diff 1.907349e-06
(x=2, y=24) diff 3.814697e-06
(x=0, y=25) diff 3.814697e-06
(x=1, y=25) diff 3.814697e-06
(x=97, y=25) diff 1.907349e-06
(x=96, y=28) diff 1.907349e-06
(x=96, y=29) diff 1.907349e-06
(x=1, y=30) diff 3.814697e-06
(x=2, y=30) diff 3.814697e-06
(x=96, y=31) diff 1.907349e-06
(x=96, y=49) diff 1.907349e-06
(x=1, y=50) diff 3.814697e-06
(x=2, y=50) diff 3.814697e-06
(x=96, y=51) diff 1.907349e-06
(x=96, y=52) diff 1.907349e-06
(x=0, y=55) diff 3.814697e-06
(x=1, y=55) diff 3.814697e-06
(x=97, y=55) diff 1.907349e-06
(x=2, y=56) diff 3.814697e-06
(x=1, y=57) diff 3.814697e-06
(x=99, y=57) diff 1.907349e-06
(x=96, y=58) diff 1.907349e-06
(x=1, y=61) diff 3.814697e-06
(x=3, y=64) diff 3.814697e-06
(x=96, y=64) diff 1.907349e-06
(x=2, y=66) diff 3.814697e-06
(x=0, y=67) diff 3.814697e-06
(x=99, y=69) diff 3.814697e-06
(x=98, y=70) diff 3.814697e-06
(x=0, y=71) diff 3.814697e-06
(x=3, y=71) diff 3.814697e-06
(x=2, y=72) diff 3.814697e-06
(x=1, y=74) diff 3.814697e-06
(x=97, y=74) diff 3.814697e-06
(x=99, y=74) diff 3.814697e-06
(x=0, y=75) diff 3.814697e-06
(x=2, y=75) diff 3.814697e-06
(x=3, y=75) diff 3.814697e-06
(x=3, y=76) diff 3.814697e-06
(x=2, y=77) diff 3.814697e-06
(x=1, y=79) diff 3.814697e-06
(x=98, y=79) diff 3.814697e-06
(x=1, y=80) diff 3.814697e-06
(x=2, y=80) diff 3.814697e-06
(x=99, y=80) diff 3.814697e-06
(x=2, y=81) diff 3.814697e-06
(x=3, y=82) diff 3.814697e-06
(x=1, y=83) diff 3.814697e-06
(x=96, y=84) diff 3.814697e-06
(x=97, y=86) diff 3.814697e-06
(x=98, y=86) diff 3.814697e-06
(x=98, y=87) diff 3.814697e-06
(x=98, y=88) diff 3.814697e-06
(x=3, y=91) diff 3.814697e-06
(x=97, y=91) diff 3.814697e-06
(x=97, y=93) diff 7.629395e-06
(x=3, y=97) diff 7.629395e-06
(x=97, y=98) diff 3.814697e-06
```

**Comment by dkurt** [comment]
This is why OpenCV calls `ippiTrueDistanceTransform_8u32f_C1R` with num threads 1 but uses own implementation otherwise:

https://github.com/opencv/opencv/blob/2c1ec4245dc793153c50fcbd255863dd1382c961/modules/imgproc/src/distransform.cpp#L780-L784

Origin of the macro is pretty old: https://github.com/opencv/opencv/commit/35c7216846ac1982e19e594e2144c11e16787c26

**Comment by dkurt** [comment]
Some analysis on non-deterministica values location:
Description - white pixels are 255, black pixels are zeros, red pixels is where `distanceTransform` returns non-deterministic values. So all non-red pixels are stable:
![img2](https://github.com/opencv/opencv/assets/25801568/4e9df1fc-b9c5-4d1a-94a7-cdf062ec7d82) ![img3](https://github.com/opencv/opencv/assets/25801568/3c013827-87bd-4203-a6b2-54134f1acebf)
![img](https://github.com/opencv/opencv/assets/25801568/d9911b79-7424-4df7-84f8-a59dc6fb1b76) ![img4](https://github.com/opencv/opencv/assets/25801568/f8473b82-a19a-4fc7-8b3d-a3ca41e60f5a)


As you may see:
1. The problematic pixels located not far from the zero pixels
2. Non-deterministic behavior reproduced only by x-axis - there is no unstable results by the y-axis

**Comment by dkurt** [comment]
After https://github.com/opencv/opencv/pull/24984 problem remains actual

**Comment by eplankin** [comment]
@dkurt, a bug to IPP was submitted.

**Comment by dkurt** [comment]
It seems to me that values `4.7683716e-07` and `9.536743e-07` that reproduced in this issue are not random small numbers. There are many artifacts in source code and GitHub issues like computational bugs: https://github.com/search?q=4.7683716e-07&type=issues

There is also a comment like that: https://github.com/osschar/mtorture/blob/c1611aee9bfcdcd1b072847485998bfd4efbf019/t4.cxx#L63-L68

---

Some other references:

> FLT_EPSILON = 9.536743e-07 The difference between 1.0 and the least value greater than 1.0 that is representable in the given floating-point type

https://www.ibm.com/docs/en/zvm/7.4?topic=descriptions-floath
